docs/design/Extension_Management.md

```diff
@@ -0,0 +1,147 @@
+
+## User Experience
+
+Extensions will now be something that are managed at a workspace level. This will allow users to have a consistent experience as well as be able to manage their extensions.
+
+Below are the commands to manage extensions:
+
+```
+upgrade-assistant extension list
+```
+
+This command will list current extensions applied to a project.
+
+```
+upgrade-assistant extension add --name [name] [--source [source]]
+```
+
+This command will allow users to add extensions to a project.
+
+```
+upgrade-assistant extension remove --name [name]
+```
+
+This command will remove an installed extension from a current workspace.
+
+```
+upgrade-assistant extension restore
+```
+
+This command will restore extensions when they are not available
+
+> This may be able to be done automatically, but initially, we'll have it be a separate command
+
+```
+upgrade-assistant extension feed list [--source [source]]
+```
+
+This command will allow someone to list extensions available on a feed
+
+### Extension Author Commands
+
+```
+upgrade-assistant extension feed publish --path [path] --source [source]
+```
+
+This command will allow extension authors to publish an extension.
+
+```
+upgrade-assistant extension feed remove --name [name] --source [source]
+```
+
+This command will allow extension authors to remove a published extension.
+
+```
+upgrade-assistant extension feed clean --source [source]
+```
+
+This command will allow extension authors to automatically clean up a feed (i.e. if a file has been removed from the storage, this will notify the author as well as remove it from the feed metadata).
+
+## Local Cache
+
+Each workspace already contains a state file that tracks progress of the tool. This state file will be updated to store extension information. This state file should be persisted into source control, which will then ensure developers working on a project will get similar results.
+
+The updated section will look similar to this:
+
+```
+{
+    ...
+    "Extensions":[
+        {
+            "Name": "...",
+            "Version": "...",
+            "Source": "...",
+            "MD5": "...",
+        }
+    ]
+}
+```
+
+This will allow `upgrade-assistant` to restore extensions from any source and ensure they are the extension that was originally installed via hash matching.
+
+When installing or restoring, packages will be retrieved from the source and expanded to disk in `%APPDATA%/dotnet/upgrade-assistant/extensions`.
+
+## Sources
+
+Below are discussed a few options for feeds we could use for distribution:
+
+### NuGet
+
+We could use NuGet as a source feed and align extension packages with with the `.nupkg` format. This would provide a number of things out of the box, such as:
+
+- Rights management for feed access
+- Versioning support
+- Validation support
+
+Problems with this approach includes:
+
+- Some extensions can be very large (such as the index files the loose dependency analysis needs)
```

Comment:
I almost feel like extension and "extension data" are perhaps 2 different concepts that deserve different thinking. Is there a reason to think of the loose dependency index as an extension itself?

Reply:
Many extensions are just data that UA will combine together (ie mapping files, template files, etc), so it follows that pattern. I'd rather keep one conceptual model of "I want to customize behavior of Upgrade Assistant". I could be convinced otherwise, though.

Reply:
Having the loose assembly indexes as a separate entity where 1 or many can associate with a loose-assembly code extension would make sense. That way if customers want to create their own private index they can just point the client code to their feed where the index is available.

Reply:
I would rather not define yet another format or deal with a separate management system. I think we can get this reconciled with one mental model. 99% of cases will be small, so the outliers we can deal with in a different way (ie host on a Azure blob using the NuGet model via [Sleet](https://www.nuget.org/packages/Sleet)

The reason I structured the loose assembly in the way I did is so that code changes wouldn't require a large download, while the index file would. The code is not feed specific; an index file is. If we start having extensions grabbing stuff from the internet, it's an additional point of internet traffic that could be difficult for some enterprises.